### PR TITLE
Fix issue 16246 - cannot call iota with 3 [u]bytes or 3 [u]shorts

### PR DESCRIPTION
**Pull request by andralex**
not much to say

### CODE REVIEW

File: std/range/package.d [hunk 1]
```diff
@@ -5747,6 +5751,18 @@ debug @system unittest
     }
 }
 
+@safe @nogc nothrow unittest
```
Comment (ZombineDev) [inline std/range/package.d]: `pure`?

File: std/range/package.d [hunk 2]
```diff
@@ -5747,6 +5751,18 @@ debug @system unittest
...
+{
+    {
+        ushort start = 0, end = 10, step = 2;
+        foreach (i; iota(start, end, step)) {}
```
Comment (ZombineDev) [inline std/range/package.d]: It may be useful to statically assert that `i` is of the expected type -`ushort` here and `ubyte` below.

File: std/range/package.d [hunk 3]
```diff
@@ -5209,6 +5209,7 @@ auto sequence(alias fun, State...)(State args)
     assert(s.front != s.front);  // no caching
 }
 
+// iota
```
Comment (schveiguy) [inline std/range/package.d]: Why this stray comment?

File: std/range/package.d [hunk 4]
```diff
@@ -5277,24 +5278,23 @@ if ((isIntegral!(CommonType!(B, E)) || isPointer!(CommonType!(B, E)))
...
+                    this.pastLast = cast(Value) (pastLast - 1);
                     this.pastLast -= (this.pastLast - current) % step;
                 }
                 else
                 {
-                    if (step < 0)
-                        assert(unsigned((current - pastLast) / -step) <= size_t.max);
-
-                    this.pastLast = pastLast + 1;
+                    assert(step == 0 || unsigned((current - pastLast) / -step) <= size_t.max);
```
Comment (schveiguy) [inline std/range/package.d]: step cannot be 0 here.

`assert(step != 0 && ...)`

File: std/range/package.d [hunk 5]
```diff
@@ -5277,24 +5278,23 @@ if ((isIntegral!(CommonType!(B, E)) || isPointer!(CommonType!(B, E)))
...
             {
                 this.step = step;
                 this.current = current;
                 if (step > 0)
                 {
                     assert(unsigned((pastLast - current) / step) <= size_t.max);
-
-                    this.pastLast = pastLast - 1;
+                    // Cast below can't fail because current < pastLast
+                    this.pastLast = cast(Value) (pastLast - 1);
```
Comment (schveiguy) [inline std/range/package.d]: This logic is correct, but hard to follow. 

I'm wondering if it wouldn't be better to refactor this into three if branches:

```D
if(current < pastLast && step > 0)
{
    this.step = step;
    this.current = current;
    ... // code currently inside the step > 0 branch
    this.pastLast += step; // this could probably be folded into last line
}
else if(current > pastLast && step < 0)
{
    this.step = step;
    this.current = current;
    ... // code currently inside the else branch of the inner if block
    this.pastLast += step;
}
else
{
    ... // step is 0 or current is on wrong side of pastLast
}
```

File: std/range/package.d (hunk 4 above)
Comment (schveiguy) [inline std/range/package.d]: Or, you can just remove that part of the assert.

File: std/range/package.d (hunk 2 above)
Comment (schveiguy) [inline std/range/package.d]: or, just `foreach(ushort i; ...)`

File: std/range/package.d (hunk 1 above)
Comment (andralex) [inline std/range/package.d]: ok

File: std/range/package.d (hunk 2 above)
Comment (andralex) [inline std/range/package.d]: ok

File: std/range/package.d (hunk 3 above)
Comment (andralex) [inline std/range/package.d]: Helps quickly locating the start of the implementation.

File: std/range/package.d (hunk 5 above)
Comment (andralex) [inline std/range/package.d]: It's a a toss seeing that this variant has more duplication. I'll try a variant with early returns.

File: std/range/package.d (hunk 4 above)
Comment (andralex) [inline std/range/package.d]: ok

File: std/range/package.d (hunk 5 above)
Comment (schveiguy) [inline std/range/package.d]: > It's a a toss seeing that this variant has more duplication.

True, but not duplication of conditionals that have already been checked :) There is also the inferred logic in your comment (if step > 0, then current must be < lastPast due to how the big if condition is constructed) that becomes unnecessary when the actual if condition is simplified.

This seems to me like the debate about combining versions and repeating the declarations. One is more DRY, but the other makes it clear which declarations are active for a specific version.

In any case, I'm OK with the code as written as it is correct and not really more confusing than the original, it just was hard to follow.

**Comment by andralex** [comment]
Had to overhaul things a bit, representing the limit as "last" instead of "past last" because of overflows.

**Comment by schveiguy** [comment]
Excellent, looks MUCH better and easier to follow the logic! Should be good to merge.

Using `last` instead of `pastLast` helps fix the issue where you want to iterate through all `ubyte`s, for instance. I don't think there's a way to actually create such a struct, however :) `iota(ubyte(0), 256)` will make the thing return `uint`s, and `iota(ubyte(0), ubyte(255))` won't go through 255. But given the way the struct is now written, it would be possible to iterate all the way through 255. I'm curious if at some future point, this would be possible (maybe a compile-time config flag?), as it's one of those things that crops up periodically.

**Comment by andralex** [comment]
Another nice thing to do with iota is have it accept a compile-time step. I.e. the `Result` structure would be parameterized by the `step` value. If the static value is zero then it's understood it's dynamic and it's a regular member etc. That way we unify a bunch of code and make it faster.
